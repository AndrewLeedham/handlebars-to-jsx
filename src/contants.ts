/**
 * Default namespace name
 */
export const DEFAULT_NAMESPACE_NAME = 'item'
