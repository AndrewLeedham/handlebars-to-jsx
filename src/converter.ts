/**
 * Utils for converting Glimmer AST (Handlebars) to Babel AST (EcmaScript)
 */
import { AST }                   from '@glimmer/syntax'
import * as Babel                from '@babel/types'
import * as isSelfClosing        from 'is-self-closing'
import * as convertHTMLAttribute from 'react-attr-converter'

/**
 * Detects expression type and converts to appropriate jsx expression
 */
export const convertExpression = (statement: AST.Statement): Babel.Expression => {
  switch (statement.type) {
    case 'ElementNode': {
      return convertElement(statement)
    }

    case 'TextNode': {
      return Babel.stringLiteral(statement.chars)
    }

    case 'MustacheStatement': {
      return convertMustacheStatement(statement)
    }

    default: {
      throw new Error('Unexpected statement')
    }
  }
}

/**
 * Converts mustache statement
 */
export const convertMustacheStatement = (statement: AST.MustacheStatement) => {
  const path = statement.path

  switch (path.type) {
    case 'PathExpression': {
      return createMemberExpression(path.parts)
    }

    case 'BooleanLiteral': {
      return Babel.booleanLiteral(path.value)
    }

    case 'NullLiteral': {
      return Babel.nullLiteral()
    }

    case 'NumberLiteral': {
      return Babel.numericLiteral(path.value)
    }

    case 'StringLiteral': {
      return Babel.stringLiteral(path.value)
    }

    case 'UndefinedLiteral': {
      return Babel.identifier('undefined')
    }

    default: {
      throw new Error('Unexpected mustache statement')
    }
  }
}

/**
 * Creates chain of member expression parts
 */
export const createMemberExpression = (parts: string[]): Babel.Identifier | Babel.MemberExpression => {
  if (parts.length === 0) {
    throw new Error('Unexpected empty expression parts')
  }

  return parts.reduce(
    (acc, item) => (acc == null ? Babel.identifier(item) : Babel.memberExpression(acc, Babel.identifier(item))),
    null as null | Babel.Identifier | Babel.MemberExpression
  )!
}

/**
 * Creates children list from js expression
 */
export const createJSXItem = (expressions: Babel.Expression[]): Babel.JSXElement['children'] =>
  expressions.map(
    expression =>
      expression.type === 'JSXElement' || expression.type === 'JSXFragment'
        ? expression
        : Babel.jsxExpressionContainer(expression)
  )

/**
 * Creates program statement
 */
export const convertProgram = (program: AST.Program) => {
  const expression
    = program.body.length === 1
      ? convertExpression(program.body[0])
      : createFragment(createJSXItem(program.body.map(statement => convertExpression(statement))))

  return Babel.program([Babel.expressionStatement(expression)])
}

/**
 * Creates JSX fragment
 */
export const createFragment = (children: Babel.JSXFragment['children']) => {
  const openingFragment = Babel.jsxOpeningFragment()
  const closingFragment = Babel.jsxClosingFragment()

  return Babel.jsxFragment(openingFragment, closingFragment, children)
}

/**
 * Creates attribute value concatenation
 */
export const createConcat = (parts: AST.ConcatStatement['parts']): Babel.BinaryExpression => {
  return parts.reduce(
    (acc, item) => {
      if (acc == null) {
        return convertExpression(item)
      }

      return Babel.binaryExpression('+', acc, convertExpression(item))
    },
    null as null | Babel.Expression | Babel.BinaryExpression
  ) as Babel.BinaryExpression
}

/**
 * Coverts AttrNode to JSXAttribute
 */
export const convertAttribute = (attrNode: AST.AttrNode): Babel.JSXAttribute | null => {
  // Unsupported attribute
  if (!/^[_\-A-z0-9]+$/.test(attrNode.name)) {
    return null
  }

  const reactAttrName = convertHTMLAttribute(attrNode.name)
  const name = Babel.jsxIdentifier(reactAttrName)
  const value = attrNode.value

  switch (value.type) {
    case 'TextNode': {
      return Babel.jsxAttribute(name, Babel.stringLiteral(value.chars))
    }

    case 'MustacheStatement': {
      return Babel.jsxAttribute(name, Babel.jsxExpressionContainer(convertMustacheStatement(value)))
    }

    case 'ConcatStatement': {
      const expression = createConcat(value.parts)

      return Babel.jsxAttribute(name, Babel.jsxExpressionContainer(expression))
    }

    default: {
      throw new Error('Unexpected attribute value')
    }
  }
}

/**
 * Converts ElementNode to JSXElement
 */
export const convertElement = (node: AST.ElementNode): Babel.JSXElement => {
  const tagName = Babel.jsxIdentifier(node.tag)
  const attributes = node.attributes.map(item => convertAttribute(item)).filter(Boolean) as Babel.JSXAttribute[]
  const isElementSelfClosing = node.selfClosing || isSelfClosing(node.tag)
  const openingElement = Babel.jsxOpeningElement(tagName, attributes, isElementSelfClosing)
  const closingElement = Babel.jsxClosingElement(tagName)

  return Babel.jsxElement(openingElement, closingElement, [], isElementSelfClosing)
}
