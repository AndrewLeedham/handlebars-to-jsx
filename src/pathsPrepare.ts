import { AST as Glimmer, traverse } from '@glimmer/syntax'
import * as hash                    from 'object-hash'
import { DEFAULT_NAMESPACE_NAME }   from './contants'

/**
 * Checks is each statement
 */
const isEachStatement = (node: Glimmer.BlockStatement) => node.path.original === 'each'

/**
 * Creates stack of namespaces
 */
const createNamespaceStack = () => {
  const namespaces: { node: Glimmer.Node; name: string; hash: string }[] = []

  return {
    // Getter of length
    get length() {
      return namespaces.length
    },

    /** Pushes sub namespaces */
    push: (item: { node: Glimmer.Node; name?: string }) =>
      namespaces.push({
        node: item.node,
        hash: hash(item.node),
        name: item.name || DEFAULT_NAMESPACE_NAME
      }),

    /** Goes to namespace up */
    pop: () => namespaces.pop(),

    /** Returns head item of the stack */
    head: () => namespaces[namespaces.length - 1]
  }
}

/**
 * Prepares paths Glimmer AST for compatible with JS AST.
 */
export const prepareProgramPaths = (program: Glimmer.Program, isComponent: boolean) => {
  const namespaces = createNamespaceStack()

  // Global component namespace
  if (isComponent) {
    namespaces.push({ node: program, name: 'props' })
  }

  traverse(program, {
    // Process block statements
    BlockStatement: {
      enter(node: Glimmer.BlockStatement) {
        // Creates new namespace to prefix all contained path expressions in
        // the block scope
        if (isEachStatement(node)) {
          namespaces.push({ node })
        }
      },
      exit(node: Glimmer.BlockStatement) {
        // Exit from namespace
        if (isEachStatement(node)) {
          namespaces.pop()
        }
      }
    },

    // Process path expressions
    PathExpression(node: Glimmer.PathExpression) {
      // Add prefixes
      if (namespaces.length) {
        node.parts.unshift(namespaces.head().name)
      }
    }
  })
}
