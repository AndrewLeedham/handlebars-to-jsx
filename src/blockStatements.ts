import { AST as Glimmer }                                                  from '@glimmer/syntax'
import * as Babel                                                          from '@babel/types'
import { resolveExpression, createRootChildren, createPath, appendToPath } from './expressions'

/**
 * Resolves block type
 */
export const resolveBlockStatement = (blockStatement: Glimmer.BlockStatement) => {
  switch (blockStatement.path.original) {
    case 'if': {
      return createConditionStatement(blockStatement)
    }

    case 'each': {
      const eachStatement = createEachStatement(blockStatement)

      // Babel.traverse(eachStatement, (node, parents) => {
      //   console.log(node, parents)
      // })

      return eachStatement
    }

    default: {
      throw new Error(`Unexpected ${blockStatement.path.original} statement`)
    }
  }
}

/**
 * Creates condition statement
 */
export const createConditionStatement = (
  blockStatement: Glimmer.BlockStatement
): Babel.ConditionalExpression | Babel.LogicalExpression => {
  const { program, inverse } = blockStatement
  const boolCondSubject: Babel.CallExpression = Babel.callExpression(
    Babel.identifier('Boolean'),
    [resolveExpression(blockStatement.params[0])] //
  )

  if (inverse == null) {
    // Logical expression
    // {Boolean(variable) && <div />}
    return Babel.logicalExpression('&&', boolCondSubject, createRootChildren(program.body))
  } else {
    // Ternary expression
    // {Boolean(variable) ? <div /> : <span />}
    return Babel.conditionalExpression(
      boolCondSubject,
      createRootChildren(program.body),
      createRootChildren(inverse.body)
    )
  }
}

/**
 * Creates each block statement
 */
export const createEachStatement = (blockStatement: Glimmer.BlockStatement) => {
  const pathExpression = blockStatement.params[0] as Glimmer.PathExpression
  const eachSubject = appendToPath(createPath(pathExpression), Babel.identifier('map'))

  const callback = Babel.arrowFunctionExpression(
    [Babel.identifier('item')],
    createRootChildren(blockStatement.program.body)
  )

  return Babel.callExpression(eachSubject, [callback])
}
