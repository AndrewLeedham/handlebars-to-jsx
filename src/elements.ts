import { AST as Glimmer }                                  from '@glimmer/syntax'
import * as Babel                                          from '@babel/types'
import * as isSelfClosing                                  from 'is-self-closing'
import * as convertHTMLAttribute                           from 'react-attr-converter'
import { createConcat, resolveExpression, createChildren } from './expressions'

/**
 * Creates JSX fragment
 */
export const createFragment = (children: Babel.JSXFragment['children']) => {
  const openingFragment = Babel.jsxOpeningFragment()
  const closingFragment = Babel.jsxClosingFragment()

  return Babel.jsxFragment(openingFragment, closingFragment, children)
}

/**
 * Coverts AttrNode to JSXAttribute
 */
export const createAttribute = (attrNode: Glimmer.AttrNode): Babel.JSXAttribute | null => {
  // Unsupported attribute
  if (!/^[_\-A-z0-9]+$/.test(attrNode.name)) {
    return null
  }

  const reactAttrName = convertHTMLAttribute(attrNode.name)
  const name = Babel.jsxIdentifier(reactAttrName)
  const value = attrNode.value

  switch (value.type) {
    case 'TextNode': {
      return Babel.jsxAttribute(name, Babel.stringLiteral(value.chars))
    }

    case 'MustacheStatement': {
      return Babel.jsxAttribute(name, Babel.jsxExpressionContainer(resolveExpression(value.path)))
    }

    case 'ConcatStatement': {
      const expression = createConcat(value.parts)

      return Babel.jsxAttribute(name, Babel.jsxExpressionContainer(expression))
    }

    default: {
      throw new Error('Unexpected attribute value')
    }
  }
}

/**
 * Converts ElementNode to JSXElement
 */
export const convertElement = (node: Glimmer.ElementNode): Babel.JSXElement => {
  const tagName = Babel.jsxIdentifier(node.tag)
  const attributes = node.attributes.map(item => createAttribute(item)).filter(Boolean) as Babel.JSXAttribute[]
  const isElementSelfClosing = node.selfClosing || isSelfClosing(node.tag)
  const children = createChildren(node.children)

  return Babel.jsxElement(
    Babel.jsxOpeningElement(tagName, attributes, isElementSelfClosing),
    Babel.jsxClosingElement(tagName),
    isElementSelfClosing ? [] : children,
    isElementSelfClosing
  )
}
