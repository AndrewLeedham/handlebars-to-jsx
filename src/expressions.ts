import { AST as Glimmer }                 from '@glimmer/syntax'
import * as Babel                         from '@babel/types'
import { createFragment, convertElement } from './elements'
import { resolveBlockStatement }          from './blockStatements'
import { createComment }                  from './comment'

/**
 * Converts the Handlebars expression to NON-JSX JS-compatible expression
 */
export const resolveStatement = (statement: Glimmer.Statement): Babel.Expression => {
  switch (statement.type) {
    case 'ElementNode': {
      return convertElement(statement)
    }

    case 'TextNode': {
      return Babel.stringLiteral(statement.chars)
    }

    case 'MustacheStatement': {
      return resolveExpression(statement.path)
    }

    case 'BlockStatement': {
      return resolveBlockStatement(statement)
    }

    case 'MustacheCommentStatement': {
      throw new Error('Top level comments currently is not supported')
    }

    default: {
      throw new Error(`Unexpected expression "${statement.type}"`)
    }
  }
}

/**
 * Converts the Handlebars node to JSX-children-compatible child element
 */
export const resolveElementChild = (
  statement: Glimmer.Statement
): Babel.JSXText | Babel.JSXElement | Babel.JSXExpressionContainer => {
  switch (statement.type) {
    case 'ElementNode': {
      return convertElement(statement)
    }

    case 'TextNode': {
      return Babel.jsxText(statement.chars)
    }

    case 'MustacheCommentStatement': {
      return createComment(statement)
    }

    // If it expression, create a expression container
    default: {
      return Babel.jsxExpressionContainer(resolveStatement(statement))
    }
  }
}

/**
 * Converts Hbs expression to Babel expression
 */
export const resolveExpression = (
  expression: Glimmer.Expression
): Babel.Literal | Babel.Identifier | Babel.MemberExpression => {
  switch (expression.type) {
    case 'PathExpression': {
      return createMemberExpression(expression.parts)
    }

    case 'BooleanLiteral': {
      return Babel.booleanLiteral(expression.value)
    }

    case 'NullLiteral': {
      return Babel.nullLiteral()
    }

    case 'NumberLiteral': {
      return Babel.numericLiteral(expression.value)
    }

    case 'StringLiteral': {
      return Babel.stringLiteral(expression.value)
    }

    case 'UndefinedLiteral': {
      return Babel.identifier('undefined')
    }

    default: {
      throw new Error('Unexpected mustache statement')
    }
  }
}

/**
 * Creates chain of member expression parts
 */
export const createMemberExpression = (parts: string[]): Babel.Identifier | Babel.MemberExpression => {
  if (parts.length === 0) {
    throw new Error('Unexpected empty expression parts')
  }

  return parts.reduce(
    (acc, item) => (acc == null ? Babel.identifier(item) : Babel.memberExpression(acc, Babel.identifier(item))),
    null as null | Babel.Identifier | Babel.MemberExpression
  )!
}

/**
 * Converts child statements of element to JSX-compatible expressions
 * @param body List of Glimmer statements
 */
export const createChildren = (body: Glimmer.Statement[]): Babel.JSXElement['children'] =>
  body.map(statement => resolveElementChild(statement))

/**
 * Converts root children
 */
export const createRootChildren = (body: Glimmer.Statement[]): Babel.Expression =>
  body.length === 1 ? resolveStatement(body[0]) : createFragment(createChildren(body))

/**
 * Creates program statement
 */
export const createProgram = (program: Glimmer.Program): Babel.Program => {
  const body = [Babel.expressionStatement(createRootChildren(program.body))]

  return Babel.program(body)
}

/**
 * Creates attribute value concatenation
 */
export const createConcat = (parts: Glimmer.ConcatStatement['parts']): Babel.BinaryExpression | Babel.Expression => {
  return parts.reduce(
    (acc, item) => {
      if (acc == null) {
        return resolveStatement(item)
      }

      return Babel.binaryExpression('+', acc, resolveStatement(item))
    },
    null as null | Babel.Expression | Babel.BinaryExpression
  ) as Babel.BinaryExpression | Babel.Expression
}
