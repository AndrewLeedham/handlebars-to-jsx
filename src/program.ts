import { AST as Glimmer }      from '@glimmer/syntax'
import * as Babel              from '@babel/types'
import { createRootChildren }  from './expressions'
import { prepareProgramPaths } from './pathsPrepare'
import { createComponent }     from './componentCreator'

/**
 * Creates program statement
 * @param hbsProgram The Handlebars program (root AST node)
 * @param isModule Should output code be exported by default
 */
export const createProgram = (hbsProgram: Glimmer.Program, isComponent: boolean, isModule: boolean, includeImport: boolean): Babel.Program => {
  prepareProgramPaths(hbsProgram, isComponent)

  const reactImport = Babel.importDeclaration([Babel.importDefaultSpecifier(Babel.identifier('React'))], Babel.stringLiteral('react'));
  const componentBody = createRootChildren(hbsProgram.body)
  const expression = isComponent ? createComponent(componentBody) : componentBody
  const statement = isModule ? Babel.exportDefaultDeclaration(expression) : Babel.expressionStatement(expression)

  const directives: Babel.Statement[] = [statement];
  includeImport && directives.unshift(reactImport);

  return Babel.program(directives)
}
