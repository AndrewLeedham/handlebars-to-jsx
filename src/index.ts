/* eslint-disable import/export */
import { preprocess }    from '@glimmer/syntax'
import generate          from '@babel/generator'
import * as Babel        from '@babel/types'
import { createProgram } from './program'

/**
 * Converts Handlebars code to JSX code
 * @param hbsCode Handlebars code to JSX
 * @param options Compilation options
 * @param [options.isComponent] Should wraps code to component
 * @param [options.isModule] Should be export code by default
 */
export function compile(hbsCode: string, isComponent?: boolean): string
export function compile(hbsCode: string, options?: { isComponent?: boolean; isModule?: boolean }): string
export function compile(
  hbsCode: string,
  options: boolean | { isComponent?: boolean; isModule?: boolean } = true
): string {
  if (typeof options === 'boolean') {
    return compile(hbsCode, { isComponent: options })
  }

  const isComponent = !!options.isComponent
  const isModule = !!options.isModule

  const glimmerProgram = preprocess(hbsCode)
  const babelProgram: Babel.Program = createProgram(glimmerProgram, isComponent, isModule)

  return generate(babelProgram).code
}
