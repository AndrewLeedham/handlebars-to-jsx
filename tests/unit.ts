import { camelizePropName, createStyleObject, parseStyleString } from '../src/styles'

describe('styles', () => {
  test('should convert style prop name to camel case', () => {
    expect(camelizePropName('class-name')).toBe('className')
    expect(camelizePropName('background-image')).toBe('backgroundImage')
    expect(camelizePropName('-webit-transition')).toBe('WebitTransition')
  })

  test('should create style object from style string', () => {
    expect(
      createStyleObject(
        "background-image: url('image.png'); margin-left: 10px;   -webit-transition: opacity 1s ease-out"
      )
    ).toEqual({
      backgroundImage: "url('image.png')",
      marginLeft:      '10px',
      WebitTransition: 'opacity 1s ease-out'
    })
  })

  test('should create ast tree of style object from style string', () => {
    expect(
      parseStyleString(
        "background-image: url('image.png'); margin-left: 10px;   -webit-transition: opacity 1s ease-out"
      )
    ).toMatchSnapshot()
  })
})
