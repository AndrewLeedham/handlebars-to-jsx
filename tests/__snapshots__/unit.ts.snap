// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styles should create ast tree of style object from style string 1`] = `
Node {
  "comments": Array [],
  "end": 98,
  "loc": SourceLocation {
    "end": Position {
      "column": 98,
      "line": 1,
    },
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "properties": Array [
    Node {
      "computed": false,
      "end": 37,
      "key": Node {
        "end": 18,
        "extra": Object {
          "raw": "\\"backgroundImage\\"",
          "rawValue": "backgroundImage",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 18,
            "line": 1,
          },
          "start": Position {
            "column": 1,
            "line": 1,
          },
        },
        "start": 1,
        "type": "StringLiteral",
        "value": "backgroundImage",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 37,
          "line": 1,
        },
        "start": Position {
          "column": 1,
          "line": 1,
        },
      },
      "method": false,
      "shorthand": false,
      "start": 1,
      "type": "ObjectProperty",
      "value": Node {
        "end": 37,
        "extra": Object {
          "raw": "\\"url('image.png')\\"",
          "rawValue": "url('image.png')",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 37,
            "line": 1,
          },
          "start": Position {
            "column": 19,
            "line": 1,
          },
        },
        "start": 19,
        "type": "StringLiteral",
        "value": "url('image.png')",
      },
    },
    Node {
      "computed": false,
      "end": 57,
      "key": Node {
        "end": 50,
        "extra": Object {
          "raw": "\\"marginLeft\\"",
          "rawValue": "marginLeft",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 50,
            "line": 1,
          },
          "start": Position {
            "column": 38,
            "line": 1,
          },
        },
        "start": 38,
        "type": "StringLiteral",
        "value": "marginLeft",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 57,
          "line": 1,
        },
        "start": Position {
          "column": 38,
          "line": 1,
        },
      },
      "method": false,
      "shorthand": false,
      "start": 38,
      "type": "ObjectProperty",
      "value": Node {
        "end": 57,
        "extra": Object {
          "raw": "\\"10px\\"",
          "rawValue": "10px",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 57,
            "line": 1,
          },
          "start": Position {
            "column": 51,
            "line": 1,
          },
        },
        "start": 51,
        "type": "StringLiteral",
        "value": "10px",
      },
    },
    Node {
      "computed": false,
      "end": 97,
      "key": Node {
        "end": 75,
        "extra": Object {
          "raw": "\\"WebitTransition\\"",
          "rawValue": "WebitTransition",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 75,
            "line": 1,
          },
          "start": Position {
            "column": 58,
            "line": 1,
          },
        },
        "start": 58,
        "type": "StringLiteral",
        "value": "WebitTransition",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 97,
          "line": 1,
        },
        "start": Position {
          "column": 58,
          "line": 1,
        },
      },
      "method": false,
      "shorthand": false,
      "start": 58,
      "type": "ObjectProperty",
      "value": Node {
        "end": 97,
        "extra": Object {
          "raw": "\\"opacity 1s ease-out\\"",
          "rawValue": "opacity 1s ease-out",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 97,
            "line": 1,
          },
          "start": Position {
            "column": 76,
            "line": 1,
          },
        },
        "start": 76,
        "type": "StringLiteral",
        "value": "opacity 1s ease-out",
      },
    },
  ],
  "start": 0,
  "type": "ObjectExpression",
}
`;
